src/v/resource_mgmt/memory_groups.h

```diff
@@ -11,48 +11,73 @@
+#include <cstddef>
 
-#include <seastar/core/memory.hh>
+/**
+ * Centralized unit for memory management.
+ *
+ * Works via a share system. First we subtract the amount of memory the user
+ * decides to use for their WebAssembly functions, then we carve out 30% for the
+ * batch cache. The remaining subsystems are distributed memory via a share
+ * system.
+ */
+class system_memory_groups {
+public:
+    system_memory_groups(size_t total_system_memory, bool wasm_enabled);
+
+    size_t kafka_total_memory() const;
 
-// centralized unit for memory management
-struct memory_groups {
-    static size_t kafka_total_memory() {
-        // 30%
-        return ss::memory::stats().total_memory() * .30;
-    }
     /// \brief includes raft & all services
-    static size_t rpc_total_memory() {
-        // 20%
-        return ss::memory::stats().total_memory() * .20;
-    }
+    size_t rpc_total_memory() const;
 
     /**
      * The target allocation size for the chunk cache. This is a soft target,
      * and may be expanded as needed by segment appenders, or reclaimed from by
      * seastar under memory pressure.
-     *
-     * add upper bound of 30 pct of memory as a hard outstanding limit.
      */
-    static size_t chunk_cache_min_memory() {
-        return ss::memory::stats().total_memory() * .10; // NOLINT
-    }
+    size_t chunk_cache_min_memory() const;
 
     /**
      * Upper bound on the amount of outstanding memory for inflight write
      * requests. Requests above this limit will wait for an existing chunk to be
      * returned to the cache.
      */
-    static size_t chunk_cache_max_memory() {
-        return ss::memory::stats().total_memory() * .30; // NOLINT
-    }
+    size_t chunk_cache_max_memory() const;
 
-    static size_t recovery_max_memory() {
-        return ss::memory::stats().total_memory() * .10;
-    }
+    size_t recovery_max_memory() const;
 
     /// Max memory that both cloud storage uploads and read-path could use
-    static size_t tiered_storage_max_memory() {
-        return ss::memory::stats().total_memory() * .10;
-    }
+    size_t tiered_storage_max_memory() const;
+
+    /// Max memory that data transform subsystem should use.
+    size_t data_transforms_max_memory() const;
+
+private:
+    /**
+     * Total memory for a core after the user's Wasm reservation.
+     */
+    size_t total_memory() const;
+    /**
+     * The amount of memory we reserve for the rest of the system to use, after
+     * we take the minimum amount required for the batch cache.
+     */
+    template<size_t shares>
+    size_t subsystem_memory() const;
+
+    size_t _total_system_memory;
+    bool _wasm_enabled;
 };
+
+/**
+ * Initialize the memory groups for **this core**.
+ *
+ * Initializing memory groups reads from cluster configuration values that can
+ * change, but require a restart to take effect, so we read and lock in the
+ * configuration values when initializing memory groups.
+ */
+void initialize_memory_groups();
```

Comment:
Could this just be done automatically inside `memory_units()` on first-call? Then if you want to init at a specific point just call `memory_groups()`?

Reply:
Fair, I initially wanted to force init to be called, but that breaks a bunch of tests, so I'll just remove this method.

Reply:
We do call this on application startup after we read the configuration to "lock in" the values.